csharp/src/Ice/InstrumentationI.cs

```diff
@@ -23,1072 +22,748 @@ public InvocationMetrics()
+                Add("requestId", obj => (obj as DispatchHelper)?.Current.RequestId);
+                Add("mode", obj => (obj as DispatchHelper)?.Mode);
             }
         }
+    }
 
-        public string GetId()
+    internal class DispatchObserver : ObserverWithDelegate<DispatchMetrics, IDispatchObserver>, IDispatchObserver
+    {
+        public void RemoteException()
         {
-            if (_id == null)
-            {
-                if (_proxy != null)
-                {
-                    var sb = new StringBuilder();
-                    try
-                    {
-                        sb.Append(_proxy.Clone(endpoints: Array.Empty<Endpoint>()));
-                        sb.Append(" [").Append(_operation).Append(']');
-                    }
-                    catch (Exception)
-                    {
-                        // Either a fixed proxy or the communicator is destroyed.
-                        sb.Append(_proxy.Identity.ToString(_proxy.Communicator.ToStringMode));
-                        sb.Append(" [").Append(_operation).Append(']');
-                    }
-                    _id = sb.ToString();
-                }
-                else
-                {
-                    _id = _operation;
-                }
-            }
-            return _id;
-        }
-
-        public string GetParent() => "Communicator";
-
-        public IObjectPrx? GetProxy() => _proxy;
-
-        public string GetEncoding() => _proxy?.Encoding.ToString() ?? "";
-
-        public string GetIdentity()
-        {
-            if (_proxy != null)
-            {
-                return _proxy.Identity.ToString(_proxy.Communicator.ToStringMode);
-            }
-            else
-            {
-                return "";
-            }
-        }
-
-        public string GetOperation() => _operation;
-
-        private readonly IObjectPrx? _proxy;
-        private readonly string _operation;
-        private readonly IReadOnlyDictionary<string, string> _context;
-        private string? _id;
-    }
-
-    internal class ThreadHelper : MetricsHelper<ThreadMetrics>
-    {
-        private class AttributeResolverI : AttributeResolver
-        {
-            public AttributeResolverI()
-            {
-                try
-                {
-                    Type cl = typeof(ThreadHelper);
-                    Add("parent", cl.GetField("Parent")!);
-                    Add("id", cl.GetField("Id")!);
-                }
-                catch (Exception)
-                {
-                    Debug.Assert(false);
-                }
-            }
-        }
-        private static readonly AttributeResolver _attributes = new AttributeResolverI();
-
-        public ThreadHelper(string parent, string id, ZeroC.Ice.Instrumentation.ThreadState state)
-            : base(_attributes)
-        {
-            Parent = parent;
-            Id = id;
-            _state = state;
+            ForEach(RemoteException);
```

Comment:
It's a bit confusing to call RemoteException from RemoteException, use a lambda here instead?